release/v1alpha1/types.go

```diff
@@ -0,0 +1,185 @@
+type ReleasePayloadSpec struct {
+	// ReleaseTag a pointer to the imagestreamtag in the TargetImageStream
+	ReleaseTag v1.ImageStreamTag `json:"releaseTag,omitempty"`
+
+	// SourceImageStream a pointer to the imagestream where this ReleasePayload was created from
+	SourceImageStream *v1.ImageStream `json:"source,omitempty"`
+
+	// TargetImageStream a pointer to the imagestream where this ReleasePayload will be located
+	TargetImageStream *v1.ImageStream `json:"target,omitempty"`
+}
+
+// ReleasePayloadStatus the status of all the promotion test jobs
+type ReleasePayloadStatus struct {
+	// Conditions communicates the state of the ReleasePayload.
+	//
+	// Supported conditions include PayloadCreated, Failed, Accepted, and Rejected.
+	//
+	// If PayloadCreated is false the ReleasePayload is waiting for a release image to be created and pushed to the
+	// TargetImageStream.  If PayloadCreated is true a release image has been created and pushed to the TargetImageStream.
+	// Verification jobs should begin and will update the status as they complete.
+	//
+	// If Failed is true a ReleasePayload image cannot be created for the given set of image mirrors
+	// This condition is terminal
+	//
+	// If Accepted is true the ReleasePayload has passed its verification criteria and can safely
+	// be promoted to an external location
+	// This condition is terminal
+	//
+	// if Rejected is true the ReleasePayload has failed one or more of its verification criteria
+	// The release-controller will take no more action in this phase, but a human may set the
+	// phase back to Ready to retry and the controller will attempt verification again.
+	Conditions []metav1.Condition `json:"conditions,omitempty"`
+
+	// BlockingJobResults stores the results of all blocking jobs
+	BlockingJobResults []JobStatus `json:"blockingJobResults,omitempty"`
+
+	// InformingJobResults stores the results of all informing jobs
+	InformingJobResults []JobStatus `json:"informingJobResults,omitempty"`
+
+	// AnalysisJobResults stores the results of all analysis jobs
+	AnalysisJobResults []JobStatus `json:"analysisJobResults,omitempty"`
+}
+
+// JobState the aggregate state of the job
+// Supported values include Pending, Failed, and Success.
+type JobState string
+
+const (
+	// JobStatePending not all job runs have completed
+	// Transitions to Failed or Success
+	JobStatePending JobState = "Pending"
+
+	// JobStateFailed failed job aggregation
+	JobStateFailed JobState = "Failed"
+
+	// JobStateSuccess successful job aggregation
+	JobStateSuccess JobState = "Success"
+)
+
+// JobStatus encapsulates the name of the job, all the results of the jobs, and an aggregated
+// result of all the jobs
+type JobStatus struct {
+	// jobName is the name of the job
+	JobName string `json:"name,omitempty"`
+
+	// MaxRetries maximum times to retry a job
+	MaxRetries int `json:"maxRetries,omitempty"`
+
+	// AggregateState is the overall success/failure of all the executed jobs
```

Comment:
tell me how this is computed.  One success means success?  One failure means failure? Something else?

Reply:
The release-controller was modified a while back to "retry" certain jobs: https://github.com/openshift/release/blob/27a01f892ad4911e37fa77fddee29e68e89a8a8f/core-services/release-controller/_releases/release-ocp-4.9.json#L27-L29

Unfortunately, this meant that the release-controller also had to track how many times that a job was retried.  To do this, it stores the number of attempts in the `release.openshift.io/verify` annotation and compares it to the `MaxRetries` of a configured job.

If we are tracking the results in `JobRunResults`, then we can calculate the number of retry attempts.  I added the `MaxRetries` here so that it's stored alongside of the individual results.

Reply:
I meant update the godoc.  What transitions are allowed for this state and how is it calculated.  I bet it's "at least one success means success", but I'd like you to say so.